jobs/db_store.go

```diff
@@ -0,0 +1,431 @@
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/weaveworks/flux"
+)
+
+// DatabaseStore is a job store backed by a sql.DB.
+type DatabaseStore struct {
+	conn   dbProxy
+	oldest time.Duration
+	now    func(dbProxy) (time.Time, error)
+}
+
+type dbProxy interface {
+	Exec(query string, args ...interface{}) (sql.Result, error)
+	Query(query string, args ...interface{}) (*sql.Rows, error)
+	QueryRow(query string, args ...interface{}) *sql.Row
+	Prepare(query string) (*sql.Stmt, error)
+}
+
+// NewDatabaseStore returns a usable DatabaseStore.
+// The DB should have a jobs table.
+func NewDatabaseStore(driver, datasource string, oldest time.Duration) (*DatabaseStore, error) {
+	conn, err := sql.Open(driver, datasource)
+	if err != nil {
+		return nil, err
+	}
+	s := &DatabaseStore{
+		conn:   conn,
+		oldest: oldest,
+		now:    nowFor(driver),
+	}
+	return s, s.sanityCheck()
+}
+
+func (s *DatabaseStore) GetJob(inst flux.InstanceID, id flux.JobID) (flux.Job, error) {
+	var (
+		queue       string
+		method      string
+		paramsBytes []byte
+		scheduledAt time.Time
+		priority    int
+		key         string
+		submittedAt time.Time
+		claimedAt   nullTime
+		heartbeatAt nullTime
+		finishedAt  nullTime
+		logStr      string
+		status      string
+		done        sql.NullBool
+		success     sql.NullBool
+	)
+	if err := s.conn.QueryRow(`
+		SELECT queue, method, params, scheduled_at, priority, key, submitted_at, claimed_at, heartbeat_at, finished_at, log, status, done, success
+		  FROM jobs
+		 WHERE id = $1
+		   AND instance_id = $2
+	`, string(id), string(inst)).Scan(
+		&queue, &method, &paramsBytes, &scheduledAt, &priority, &key, &submittedAt,
+		&claimedAt, &heartbeatAt, &finishedAt, &logStr, &status, &done, &success,
+	); err == sql.ErrNoRows {
+		return flux.Job{}, flux.ErrNoSuchJob
+	} else if err != nil {
+		return flux.Job{}, errors.Wrap(err, "error getting job")
+	}
+
+	params, err := s.scanParams(method, paramsBytes)
+	if err != nil {
+		return flux.Job{}, errors.Wrap(err, "unmarshaling params")
+	}
+
+	var log []string
+	if err := json.NewDecoder(strings.NewReader(logStr)).Decode(&log); err != nil {
+		return flux.Job{}, errors.Wrap(err, "unmarshaling log")
+	}
+
+	return flux.Job{
+		Instance:    inst,
+		ID:          id,
+		Queue:       queue,
+		Method:      method,
+		Params:      params,
+		ScheduledAt: scheduledAt,
+		Priority:    priority,
+		Key:         key,
+		Submitted:   submittedAt,
+		Claimed:     claimedAt.Time,
+		Heartbeat:   heartbeatAt.Time,
+		Finished:    finishedAt.Time,
+		Log:         log,
+		Status:      status,
+		Done:        done.Bool,
+		Success:     success.Bool,
+	}, nil
+}
+
+// PutJob schedules a job to run. Users should set the Queue, Method, Params,
+// and ScheduledAt fields of the job. If ScheduledAt is nil, the job will run
+// immediately.
+func (s *DatabaseStore) PutJob(inst flux.InstanceID, job flux.Job) (flux.JobID, error) {
```

Comment:
When trying this out, I noticed that consistently the release job did not run until just about a minute after being posted.

```
ts=2016-12-02T12:35:15Z caller=transport.go:733 method=PostRelease url="/v4/release?image=[...]
ts=2016-12-02T12:35:16Z caller=transport.go:733 method=GetRelease url="/v4/release?id=727c1b67-83c0-1b7b-747c-7b67cc1acddf" took=605.313µs status_code=200
# ... many more of those
ts=2016-12-02T12:36:03Z caller=releaser.go:349 instanceID=1 description="Release quay.io/weaveworks/helloworld:master-a000001 to default/helloworld"
# ... job actually proceeds